lib/token_scanner.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require 'json'
+require 'jwt'
+require 'base64'
+
+# TokenScanner scans configmaps for JWT-style tokens and logs approaching or expired number of days
+class TokenScanner
+  # warn threshold in days
+  WARN_THRESHOLD = 30
+
+  # Regex pattern to match JWT like strings:
+  #   with non-capturing group, capture any word character (\w) and hyphens followed by . exactly 2 times
+  #   then, capture any word character (\w) and hyphens until end of line
+  #   like: xxx.xxx.xxx, -xx.x-x-x.xx-, etc...
+  JWT_REGEX = /^(?:[\w-]*\.){2}[\w-]*$/.freeze
+
+  def initialize
+    @tokens = {
+      production: [],
+      staging: []
+    }
+  end
+
+  # run runs the scan and logs any existing results existing with an error
+  def run
+    @tokens.each do |context, results|
+      scan(context, results)
+    end
+
+    return unless @tokens[:production].any? || @tokens[:staging].any?
+
+    # log if any tokens are near (or past) expiration
+    log
+    raise 'Some tokens will expire soon or have expired!'
+  end
+
+  private
+
+  # scan scans the configmaps for expiring JWTs
+  def scan(context, results)
+    config_maps(context)['items'].each do |config_map|
+      name = config_map['metadata']['name']
+      next unless config_map.key?('data')
+
+      config_map['data'].each do |key, value|
+        next unless jwt?(value)
+
+        decoded_jwt = JWT.decode(value, nil, false)
+        next unless decoded_jwt[0].key?('exp')
+
+        days_left = get_days_left(decoded_jwt[0]['exp'])
+        results << [name, key, days_left] if days_left < WARN_THRESHOLD
+      end
+    end
+  end
+
+  def config_maps(context)
+    JSON.parse(`kubectl --context #{context} get configmaps -o json`)
+  end
+
+  # get_days_left returns the number of days left until expiration from today
+  def get_days_left(exp)
+    exp_time = Time.at(exp).utc
+    current_time = Time.now.utc + (60 * 60 * 24)
+    ((exp_time - current_time) / (60 * 60 * 24)).to_i
+  end
+
+  # jwt? returns true if the input is a JWT-style token
+  def jwt?(value)
+    return unless JWT_REGEX.match?(value)
+    # to further validate JWT, split the value on . and test the header (first part in JWT, before the first period)
+    #   JWT header should be base64 encoded and contain an 'alg' and 'typ' keys
+    #   some JWTs may not have 'typ' so we test header for 'alg' instead
+    return unless Base64.decode64(value.split('.').first).include?('alg')
```

Comment:
Might be good for this method to have a `rescue` that returns `false`, in case of JWT-looking strings that don't actually decode successfully.

Reply:
Switched to `strict_decode64` and added a `rescue ArgumentError`. Also changed to return `false` explicitly.